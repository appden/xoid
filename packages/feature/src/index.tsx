type Config<T, U> = {
  id?: string | symbol
  defaultOptions?: T
  options?: U
}

export type From = {
  <U extends new (...args: any) => any>(item: U): InstanceType<U>
  <U extends { id: any; type: any }>(item: U): U['type']
}

export abstract class FeatureAbstract<Options, DefaultOptions> {
  externalOptions!: Options
  options!: DefaultOptions & Options
  from!: From
  main?: () => void
}

export type FeatureConstructor = new (options: any) => FeatureAbstract<any, any>

type PrefilledFeature<T> = T extends new (options: any) => infer Q
  ? new (options: {}) => Omit<Q, 'externalOptions'> & { externalOptions: {} }
  : never

export type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never

type GetKey<I, K extends string | symbol | number> = I extends {
  [L in K]: infer P
}
  ? P
  : never

export type ExtractKey<U, K extends string | number | symbol> = GetKey<UnionToIntersection<U>, K>

type MergedTypes<P extends FeatureConstructor> = {
  [K in keyof UnionToIntersection<InstanceType<P>>]: ExtractKey<InstanceType<P>, K>
}

const REFERENCE = Symbol()

const defaultFrom: any = () => console.warn('Please use it inside compose')

const current = { from: defaultFrom }

const createFrom = (contextMap: any) =>
  function (this: any, input: any) {
    const output = contextMap.get(typeof input === 'object' ? input.id : input)
    if (!output) {
      console.error(`Dependency '${input.name}' not found in '${this.constructor.name}'.`)
    }
    return output
  }

const interceptFrom = <T,>(nextFrom: any, fn: () => T) => {
  current.from = nextFrom
  const ans = fn()
  current.from = defaultFrom
  return ans
}

export function Feature<T extends Config<any, any> = { defaultOptions: {}; options: {} }>(
  config = {
    id: Symbol(),
    defaultOptions: {},
    options: {},
  } as T
) {
  return class {
    static id = config.id

    externalOptions!: T['options']
    options!: T['options'] & T['defaultOptions']
    from!: From

    constructor(options: T['options']) {
      this.from = (a: any) => current.from(a)
      this.options = { ...config.defaultOptions, ...options }
    }
  }
}

export const composeFeatures =
  <P extends FeatureConstructor, V>(
    config: P[],
    afterAll: (from: From, common: MergedTypes<P>) => V
  ) =>
  (options: ExtractKey<InstanceType<P>, 'externalOptions'>) => {
    const contextMap = new Map<P, InstanceType<P>>()
    config
      .map((item) => {
        const instance = runInner(contextMap)(item, options)
        // @ts-ignore
        contextMap.set(item[REFERENCE] || item, instance)
        // @ts-ignore
        if (item.id) {
          // @ts-ignore
          contextMap.set(item.id, instance)
        }
        return instance
      })
      .forEach((instance) => {
        instance?.main?.()
      })
    const ans = afterAll(contextMap.get.bind(contextMap) as From, {} as MergedTypes<P>)
    return ans
  }

const runInner =
  (contextMap: Map<FeatureConstructor, FeatureAbstract<any, any>>) =>
  <P extends FeatureConstructor>(
    item: P,
    options: ExtractKey<InstanceType<P>, 'externalOptions'>
  ) => {
    const from = createFrom(contextMap)
    const instance = interceptFrom(from, () => new item(options))
    instance.from = from
    return instance
  }

export const prefill = <
  T extends FeatureConstructor,
  U extends ExtractKey<InstanceType<T>, 'externalOptions'>
>(
  item: T,
  options: U | ((from: From) => U)
) =>
  // @ts-ignore
  class extends item {
    static [REFERENCE] = item
    constructor() {
      // @ts-ignore
      super(typeof options === 'function' ? options(current.from) : options)
    }
  } as PrefilledFeature<T>
