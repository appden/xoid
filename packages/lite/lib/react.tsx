import { useReducer, useEffect, useLayoutEffect, useRef } from 'react'
import { create, subscribe, select } from './vanilla'
import { Observable } from '@xoid/engine'

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
const useIsoLayoutEffect = typeof window === 'undefined' ? useEffect : useLayoutEffect

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}

/**
 * Subscribes to a store, or a value inside a React function component.
 * @see [xoid.dev/docs/api-react/usestore](https://xoid.dev/docs/api-react/usestore)
 */

export function useStore<T>(store: Observable<T>): T
export function useStore<T, U>(store: Observable<T>, selector: (state: T) => U): U
export function useStore<T, U>(store: Observable<T>, selector?: (state: T) => U): any {
  const forceUpdate = useReducer((c) => c + 1, 0)[1]
  const item = useConstant(() => {
    return selector ? select(store, selector) : store
  })

  useIsoLayoutEffect(() => subscribe(item, forceUpdate), [])
  return item()
}

/**
 * Subscribes to a store, or a value inside a React function component.
 * @see [xoid.dev/docs/api-react/usesetup](https://xoid.dev/docs/api-react/usesetup)
 */

export function useSetup<T>(
  model: (deps: Observable<undefined>, onCleanup: (fn: () => void) => void) => T
): T
export function useSetup<T, P>(
  model: (deps: Observable<P>, onCleanup: (fn: () => void) => void) => T,
  props: P
): T
export function useSetup(model: (deps: any, onCleanup: any) => any, props?: any): any {
  const setup = useConstant(() => {
    const deps = create(props)
    const fns: any[] = []
    const onCleanup = (fn: any) => fns.push(fn)
    const main = model(deps, onCleanup)
    return { main, deps, fns }
  })
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useIsoLayoutEffect(() => setup.deps(props), [props])
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useIsoLayoutEffect(() => () => setup.fns.forEach((fn) => fn()), [])
  return setup.main
}
