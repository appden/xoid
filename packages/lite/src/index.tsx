export { create } from '../../xoid/src/internal/lite'
