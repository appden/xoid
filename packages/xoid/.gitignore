types
index.js
index.esm.js