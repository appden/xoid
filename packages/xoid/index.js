'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var C = require('@xoid/core');
var M = require('@xoid/model');
var R = require('@xoid/ready');
var x = Object.assign(M.model.bind({}), C, M, R);

function make(O) {
	Object.keys(O).forEach(function (k) {
		Object.defineProperty(exports, k, {
			enumerable: true,
			get: function () {
				return O[k];
			}
		});
	});
}
make(C)
make(M)
make(R)
exports.default = x;