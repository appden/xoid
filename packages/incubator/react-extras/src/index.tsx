import { useAtom, useSetup } from '@xoid/react'
export { slice } from './slice'

/**
 * Can be used to consume a **value-onChange pair** as an atom inside React components.
 *
 * Third argument, when set to `true`, enables optimistic updates. This can be
 * used for keeping a synced local state, and debouncing the updates sent to the parent.
 * @see [xoid.dev/docs/api-react/use-thru](https://xoid.dev/docs/api-react/use-thru)
 */
export const useThru = <T,>(value: T, onChange?: (value: T) => void, optimistic = false) => {
  const atom = useSetup(
    ($props) => {
      const atom = $props.focus('value')
      const props = $props.value
      const prevSet = atom.set
      atom.set = (value: T) => {
        if (props.optimistic) prevSet(value)
        props.onChange?.(value)
      }
      return atom
    },
    { value, onChange, optimistic }
  )
  useAtom(atom)
  return atom
}
