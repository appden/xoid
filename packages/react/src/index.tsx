import { useSyncExternalStore } from 'use-sync-external-store/shim'
import React, { useEffect, useLayoutEffect, useRef, useDebugValue } from 'react'
import { Atom, Usable, create, use as _use } from 'xoid'

export type Pair<T> = { value: T; onChange: (value: T) => void }

export type ComponentType = keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>

export type PropsOf<T> = T extends ComponentType
  ? React.ComponentProps<T>
  : T extends (...args: any) => any
  ? Parameters<T>[0]
  : never

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
export const useIsoLayoutEffect = window === undefined ? useEffect : useLayoutEffect

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}

const useReactAdapter = () => {
  const result = useConstant(() => {
    const mount = (_use as any).createEvent()
    const unmount = (_use as any).createEvent()
    return {
      mount,
      unmount,
      adapter: {
        mount: mount.add,
        unmount: unmount.add,
      },
    }
  })
  useIsoLayoutEffect(() => {
    result.mount.fire()
    return result.unmount.fire
  }, [])
  return result.adapter
}

export type ReactAdapter = ReturnType<typeof useReactAdapter>

/**
 * Subscribes to an atom inside a React function component.
 *
 * An atom, or a function returning an atom can be passed as the first argument.
 * When the second optional argument is set to `true`, it will also consume the usables of the atom.
 * @see [xoid.dev/docs/api-react/use-atom](https://xoid.dev/docs/api-react/use-atom)
 */
export function useAtom<T>(atom: Atom<T>): T
export function useAtom<T>(atom: () => Atom<T>): T
export function useAtom<T, U>(atom: Atom<T> & Usable<U>, use: true): [T, U]
export function useAtom<T, U>(atom: () => Atom<T> & Usable<U>, use: true): [T, U]
export function useAtom<T, U>(maybeAtom: Atom<T> | (() => Atom<T>), use?: boolean): [T, U] | T {
  const atom = useConstant(() => (typeof maybeAtom === 'function' ? maybeAtom() : maybeAtom))
  const value = useSyncExternalStore(
    atom.subscribe,
    () => atom.value,
    () => atom.value
  )
  useDebugValue(value)
  return use ? ([value, _use(atom as any)] as [T, U]) : (value as T)
}

/**
 * Can be used to create local state inside React components. Similar to `React.useMemo`,
 * but creates values **exactly once**.
 * @see [xoid.dev/docs/api-react/use-setup](https://xoid.dev/docs/api-react/use-setup)
 */
export function useSetup<T>(fn: ($props: undefined, adapter: ReactAdapter) => T): T
export function useSetup<T, P>(fn: ($props: Atom<P>, adapter: ReactAdapter) => T, props: P): T
export function useSetup(fn: ($props: any, adapter: any) => any, props?: any): any {
  /* eslint-disable react-hooks/rules-of-hooks */
  // Calling hooks conditionally wouldn't be an issue here, because we rely on just
  // the Function.length and Arguments.length, which will remain static.
  const api = fn.length > 1 ? useReactAdapter() : undefined
  if (arguments.length > 1) {
    const [$props, result] = useConstant(() => {
      const $props = create(() => props)
      const result = fn($props, api)
      return [$props, result] as const
    })
    useIsoLayoutEffect(() => ($props as Atom<any>).set(props), [props])
    return result
  } else {
    const result = useConstant(() => fn(undefined, api))
    return result
  }
  /* eslint-enable react-hooks/rules-of-hooks */
}
