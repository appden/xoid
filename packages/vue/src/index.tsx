import { getCurrentScope, onScopeDispose, shallowRef } from 'vue'
import { Atom } from 'xoid'

export function useAtom<T>(atom: Atom<T>) {
  const state = shallowRef()

  const unsubscribe = atom.subscribe((value) => {
    state.value = value
  })

  getCurrentScope() && onScopeDispose(unsubscribe)
  return state
}
