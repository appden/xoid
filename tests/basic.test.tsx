import React from 'react'
import ReactDOM from 'react-dom'
import { act, cleanup, fireEvent, render } from '@testing-library/react'
import { create, use, useStore, Store } from '../src/xoid'

const debug = (store: Store<any, any>) => {
  return {
    self: store,
    selfSerialized: JSON.stringify(store),
    get: store(),
    getSerialized: JSON.stringify(store()),
    use: use(store),
  }
}

const consoleError = console.error
afterEach(() => {
  cleanup()
  console.error = consoleError
})

it('creates a store with a primitive value', () => {
  const store = create(5)
  expect(debug(store)).toMatchSnapshot()
})

it('creates a store with a record', () => {
  const store = create({ alpha: 3, beta: 5 })
  expect(debug(store)).toMatchSnapshot()
})

it('normalizes nested stores', () => {
  const store = create(create(5))
  expect(debug(store)).toMatchSnapshot()
})

it('normalizes nested stores 2', () => {
  const store = create(create({ a: create(5), b: create(7) }))
  expect(debug(store)).toMatchSnapshot()
})

it('normalizes nested stores in a record', () => {
  const store = create({ alpha: create(3), beta: create(5) })
  expect(debug(store)).toMatchSnapshot()
})

// it('creates a selector store', () => {
//   const a = create(3)
//   const b = create(5)
//   const store = create((get) => get(a) + get(b))
//   expect(debug(store)).toMatchSnapshot()
// })

it('uses the actions in vanilla', async () => {
  const store = create({ count: 0 }, (store) => ({
    inc: () => store((state) => ({ count: state.count + 1 })),
  }))
  use(store).inc()
  expect(debug(store)).toMatchSnapshot()
})

it('uses the actions in React', async () => {
  const store = create({ count: 0 }, (store) => ({
    inc: () => store((state) => ({ count: state.count + 1 })),
  }))

  function Counter() {
    const { count } = useStore(store)
    const { inc } = use(store)
    React.useEffect(inc, [inc])
    return <div>count: {count}</div>
  }

  const { findByText } = render(<Counter />)

  await findByText('count: 1')
})

it('only runs when partial state changes in React', async () => {
  const store = create({ count: 0, count2: 'constant' }, (store) => ({
    inc: () => store((state) => ({ ...state, count: state.count + 1 })),
  }))

  let renderCount = 0

  function Counter() {
    const c2 = useStore(store.count2)
    React.useEffect(use(store).inc, [])
    renderCount++
    return <div>count: {c2}</div>
  }

  const { findByText } = render(<Counter />)

  await findByText('count: constant')
  expect(renderCount).toBe(1)
})

it('can batch updates', async () => {
  const store = create(
    {
      count: 0,
    },
    (store) => ({
      inc: () => store((state) => ({ count: state.count + 1 })),
    })
  )

  function Counter() {
    const { count } = useStore(store)
    const { inc } = use(store)
    React.useEffect(() => {
      ReactDOM.unstable_batchedUpdates(() => {
        inc()
        inc()
      })
    }, [])
    return <div>count: {count}</div>
  }

  const { findByText } = render(<Counter />)

  await findByText('count: 2')
})

it('can update the selector', async () => {
  const store = create(() => ({
    one: 'one',
    two: 'two',
  }))

  function Component({ selector }: any) {
    const [value] = useStore(selector)
    return <div>{value}</div>
  }

  const { findByText, rerender } = render(<Component selector={store.one} />)
  await findByText('one')

  // rerender(<Component selector={store.two} />)
  // await findByText('two')
})

it('can get the store', () => {
  const store = create({
    value: 1,
  })

  expect(store().value).toBe(1)
  expect(store.value()).toBe(1)
})

it('can set the store', () => {
  const store = create({
    value: 1,
  })

  store({ value: 2 })
  expect(store.value()).toBe(2)
  store({ value: 3 })
  expect(store.value()).toBe(3)
  store((state) => ({ value: state.value + 1 }))
  expect(store.value()).toBe(4)
  store.value((state) => state + 1)
  expect(store.value()).toBe(5)
})

it('ensures parent components subscribe before children', async () => {
  const store = create(() => ({
    children: {
      '1': { text: 'child 1' },
      '2': { text: 'child 2' },
    },
  }))

  function changeState() {
    set(store, {
      children: {
        '3': { text: 'child 3' },
      },
    })
  }

  function Child({ id }: any) {
    const [text] = useStore(store.children[id].text)
    return <div>{text}</div>
  }

  function Parent() {
    const [childStates] = useStore(store.children)
    return (
      <>
        <button onClick={changeState}>change state</button>
        {Object.keys(childStates).map((id: any) => (
          <Child id={id} key={id} />
        ))}
      </>
    )
  }

  const { getByText, findByText } = render(<Parent />)

  fireEvent.click(getByText('change state'))

  await findByText('child 3')
})

// https://github.com/pmndrs/zustand/issues/84
it('ensures the correct subscriber is removed on unmount', async () => {
  const store = create({ count: 0 })

  function increment() {
    store(({ count }) => ({ count: count + 1 }))
  }

  function Count() {
    const c = useStore(store.count)
    return <div>count: {c}</div>
  }

  function CountWithInitialIncrement() {
    React.useLayoutEffect(increment, [])
    return <Count />
  }

  function Component() {
    const [Counter, setCounter] = React.useState(() => CountWithInitialIncrement)
    React.useLayoutEffect(() => {
      setCounter(() => Count)
    }, [])
    return (
      <>
        <Counter />
        <Count />
      </>
    )
  }

  const { findAllByText } = render(<Component />)

  expect((await findAllByText('count: 1')).length).toBe(2)

  act(increment)

  expect((await findAllByText('count: 2')).length).toBe(2)
})

// https://github.com/pmndrs/zustand/issues/86
it('ensures a subscriber is not mistakenly overwritten', async () => {
  const store = create({ count: 0 })

  function Count1() {
    const c = useStore(store.count)
    return <div>count1: {c}</div>
  }

  function Count2() {
    const c = useStore(store.count)
    return <div>count2: {c}</div>
  }

  // Add 1st subscriber.
  const { findAllByText, rerender } = render(<Count1 />)

  // Replace 1st subscriber with another.
  rerender(<Count2 />)

  // Add 2 additional subscribers.
  rerender(
    <>
      <Count2 />
      <Count1 />
      <Count1 />
    </>
  )

  // Call all subscribers
  act(() => store({ count: 1 }))

  expect((await findAllByText('count1: 1')).length).toBe(2)
  expect((await findAllByText('count2: 1')).length).toBe(1)
})
