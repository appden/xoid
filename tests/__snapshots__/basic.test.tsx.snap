// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a atom with a primitive value 1`] = `
Object {
  "get": 5,
  "getSerialized": "5",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": 5,
    "watch": [Function],
    Symbol(): Object {
      "atom": [Circular],
      "get": [Function],
      "isStream": false,
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":5}",
  "use": undefined,
}
`;

exports[`creates a atom with a record 1`] = `
Object {
  "get": Object {
    "alpha": 3,
    "beta": 5,
  },
  "getSerialized": "{\\"alpha\\":3,\\"beta\\":5}",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": Object {
      "alpha": 3,
      "beta": 5,
    },
    "watch": [Function],
    Symbol(): Object {
      "atom": [Circular],
      "get": [Function],
      "isStream": false,
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":{\\"alpha\\":3,\\"beta\\":5}}",
  "use": undefined,
}
`;

exports[`normalizes nested atoms in a record 1`] = `
Object {
  "get": Object {
    "alpha": Object {
      "focus": [Function],
      "map": [Function],
      "set": [Function],
      "subscribe": [Function],
      "update": [Function],
      "value": 3,
      "watch": [Function],
      Symbol(): Object {
        "atom": [Circular],
        "get": [Function],
        "isStream": false,
        "listeners": Set {},
        "set": [Function],
        "subscribe": [Function],
        "usable": undefined,
      },
    },
    "beta": Object {
      "focus": [Function],
      "map": [Function],
      "set": [Function],
      "subscribe": [Function],
      "update": [Function],
      "value": 5,
      "watch": [Function],
      Symbol(): Object {
        "atom": [Circular],
        "get": [Function],
        "isStream": false,
        "listeners": Set {},
        "set": [Function],
        "subscribe": [Function],
        "usable": undefined,
      },
    },
  },
  "getSerialized": "{\\"alpha\\":{\\"value\\":3},\\"beta\\":{\\"value\\":5}}",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": Object {
      "alpha": Object {
        "focus": [Function],
        "map": [Function],
        "set": [Function],
        "subscribe": [Function],
        "update": [Function],
        "value": 3,
        "watch": [Function],
        Symbol(): Object {
          "atom": [Circular],
          "get": [Function],
          "isStream": false,
          "listeners": Set {},
          "set": [Function],
          "subscribe": [Function],
          "usable": undefined,
        },
      },
      "beta": Object {
        "focus": [Function],
        "map": [Function],
        "set": [Function],
        "subscribe": [Function],
        "update": [Function],
        "value": 5,
        "watch": [Function],
        Symbol(): Object {
          "atom": [Circular],
          "get": [Function],
          "isStream": false,
          "listeners": Set {},
          "set": [Function],
          "subscribe": [Function],
          "usable": undefined,
        },
      },
    },
    "watch": [Function],
    Symbol(): Object {
      "atom": [Circular],
      "get": [Function],
      "isStream": false,
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": undefined,
    },
  },
  "selfSerialized": "{\\"value\\":{\\"alpha\\":{\\"value\\":3},\\"beta\\":{\\"value\\":5}}}",
  "use": undefined,
}
`;
